src/System.Net.Http/src/System/Net/Http/Unix/CurlResponseMessage.cs

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.Threading;
+
+namespace System.Net.Http
+{
+    internal sealed class CurlResponseMessage : HttpResponseMessage
+    {
+        private volatile bool _disposed = false;
+        private readonly CurlResponseStream _responseStream = new CurlResponseStream();
+        private readonly ManualResetEvent _headersReceivedEvent = new ManualResetEvent(false);
+
+        #region Properties
+
+        internal ReaderWriterStream ContentStream
+        {
+            get
+            {
+                return _responseStream.InnerStream;
+            }
+        }
+
+        #endregion
+
+        internal CurlResponseMessage(HttpRequestMessage request)
+        {
+            RequestMessage = request;
+            Content = new StreamContent(_responseStream);
+        }
+
+        internal void WaitForHeaders(CancellationToken cancellationToken)
+        {
+            CheckDisposed();
+            var events = new WaitHandle[] {_headersReceivedEvent, cancellationToken.WaitHandle};
+            WaitHandle.WaitAny(events, Timeout.InfiniteTimeSpan);
+        }
+
+        internal void SignalHeadersReceived()
+        {
+            if (!_disposed)
```

Comment:
The containing object (CurlResponseMessage) will not be available to the caller of CurlHandler.SendAsync till the headers have been received and the TCS task was marked complete. Is there some other code path that could trigger Dispose?